content/1_docs/2_cookbook/9_setup/0_understanding-oop/recipe.txt

```diff
@@ -0,0 +1,468 @@
+Title: A very brief intro to object oriented programming in PHP
+
+----
+
+Published: 2020-11-09
+
+----
+
+Description: Understanding classes, objects, member functions and common pitfalls
+
+----
+
+Text:
+
+## Intro
+
+The Kirby core is written in PHP, a scripting language that still powers a majority of websites.
+
+In the past, PHP was mostly written using a procedural approach, with functions that act on data structures.
+
+Today, professional PHP programming is based on an OOP (Object Oriented Programming) architecture which has many advantages over the procedural approach. The idea behind OOP is basically to encapsulate data and the methods that act on it, in order to produce modular, extendable and maintainable software.
+
+This recipe will give you just enough background to understand Kirby's architecture and after all, to understand typical errors thrown in your code, like "calling a member function x on null" or why you can use variables like `$kirby` or `$page` in some contexts but not in others.
+
+## Prerequisites
+
+If you want to follow along with this recipe, you need the following:
+
+- A web server (ideally on your (link: docs/cookbook/setup/development-environment text: local machine))
+- A current version of PHP
+- A Kirby installation ((link: docs/guide/quickstart text: Starterkit or Plainkit))
+- A code/text editor of your choice
+
+## What is OOP?
+
+Object Oriented Programming is a programming paradigm based on the concept of objects that have certain properties and methods that allow interacting with these objects. The properties and methods of such an object are described in a class, which acts as the model or blueprint for such objects, whereas an object is an instance of this class, i.e. a specific embodiment of the class.
```

Comment:
I know there's an example right after this, but I feel this is can be too abstract for readers without OOP experience. It could be that readers stop reading because they didn't understand the explanation.

What if we first had a relatable real-world example and then the abstract explanation afterwards. Something like:

> Object Oriented Programming (OOP) is all about objects, their properties and their methods.
>
> Let's first think about a real-world example: An object could be an apple. Its properties are the variety, color, weight and size. And its methods are the actions you could do with the apple, e.g. you can pick it up, give it to someone or eat it.
>
> OOP uses the same paradigms for data structures in a program. The properties and methods of an object like the apple are described in a class, which acts as the model or blueprint for every apple you might have in your application, whereas an object is an instance of this class, i.e. a specific embodiment of the class (the specific apple you can hold in your hand).

Reply:
Great idea, thanks! Getting too abstract sometimes...
I think I'd rather stay with the book analogy though, to keep it consistent, instead of introducing apples ;)

Reply:
You are right, otherwise it really is an apple to books comparison. 😆